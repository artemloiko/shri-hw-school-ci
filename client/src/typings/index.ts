export * from '../../../server/typings/storage.interfaces';
