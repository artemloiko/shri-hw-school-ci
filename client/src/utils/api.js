import axios from 'axios';

const axiosAPI = axios.create({
  baseURL: '/api',
});

class API {
  constructor(axiosInstance) {
    this.axiosInstance = axiosInstance;
  }

  async getSettings() {
    const response = await this.axiosInstance.get('/settings');
    return response.data?.data;
  }

  /**
   * @param {object} settingsDTO School CI settings
   * @param {string} settingsDTO.repoName
   * @param {string} settingsDTO.buildCommand
   * @param {string} settingsDTO.mainBranch
   * @param {number} settingsDTO.period
   */
  async setSettings(settingsDTO) {
    const response = await this.axiosInstance.post('/settings', settingsDTO);
    return response.data?.data;
  }

  async getBuildsList(offset = 0, limit = 25) {
    const response = await this.axiosInstance.get(`/builds?offset=${offset}&limit=${limit}`);
    return response.data?.data;
  }

  async addBuild(commitHash) {
    const response = await this.axiosInstance.post(`/builds/${commitHash}`);
    return response.data?.data;
  }

  async getBuildDetails(buildId) {
    const response = await this.axiosInstance.get(`/builds/${buildId}`);
    return response.data?.data;
  }

  async getBuildLog(buildId) {
    const response = await this.axiosInstance.get(`/builds/${buildId}/logs`);
    return response.data;
  }
}

const ApiInstance = new API(axiosAPI);

export default ApiInstance;
