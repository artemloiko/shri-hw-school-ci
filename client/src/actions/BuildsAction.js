import api from '../utils/api';
import { createAction } from '../utils/createAction';
export const GET_BUILDS_LIST_REQUEST = 'GET_BUILDS_LIST_REQUEST';
export const GET_BUILDS_LIST_SUCCESS = 'GET_BUILDS_LIST_SUCCESS';
export const GET_BUILDS_LIST_FAIL = 'GET_BUILDS_LIST_FAIL';
export const GET_BUILDS_LIST_UPDATE = 'GET_BUILDS_LIST_UPDATE';

export const LOAD_MORE_BUILDS_REQUEST = 'LOAD_MORE_BUILDS_REQUEST';
export const LOAD_MORE_BUILDS_SUCCESS = 'LOAD_MORE_BUILDS_SUCCESS';
export const LOAD_MORE_BUILDS_FAIL = 'LOAD_MORE_BUILDS_FAIL';

const getBuildsListRequest = () => {
  return createAction(GET_BUILDS_LIST_REQUEST);
};
const getBuildsListSuccess = (data) => {
  return createAction(GET_BUILDS_LIST_SUCCESS, data);
};
const getBuildsListFail = (error) => {
  return createAction(GET_BUILDS_LIST_FAIL, error, true);
};

function fetchBuildsList() {
  return async (dispatch) => {
    dispatch(getBuildsListRequest());
    try {
      const data = await api.getBuildsList();
      dispatch(getBuildsListSuccess(data));
    } catch (error) {
      dispatch(getBuildsListFail(error.message));
    }
  };
}

function shouldFetchBuildsList(state) {
  const { builds } = state;

  if (builds?.isFetching || builds?.isLoaded) {
    return false;
  }

  return true;
}

export function fetchBuildsListIfNeeded() {
  return (dispatch, getState) => {
    if (shouldFetchBuildsList(getState())) {
      dispatch(fetchBuildsList());
    }
  };
}

export function updateBuildsList(data) {
  return createAction(GET_BUILDS_LIST_UPDATE, data);
}

const loadMoreBuildsRequest = () => {
  return createAction(LOAD_MORE_BUILDS_REQUEST);
};
const loadMoreBuildsSuccess = (data) => {
  return createAction(LOAD_MORE_BUILDS_SUCCESS, data);
};
const loadMoreBuildsFail = (error) => {
  return createAction(LOAD_MORE_BUILDS_FAIL, error, true);
};

export function fetchMoreBuilds() {
  return async (dispatch, getState) => {
    dispatch(loadMoreBuildsRequest());
    try {
      const { buildsList } = getState().builds;
      const data = await api.getBuildsList(buildsList.length);
      dispatch(loadMoreBuildsSuccess(data));
    } catch (error) {
      dispatch(loadMoreBuildsFail(error.message));
    }
  };
}

export async function addBuild(commitHash) {
  return api.addBuild(commitHash);
}
