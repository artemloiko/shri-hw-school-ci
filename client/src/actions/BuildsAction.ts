import api from '../utils/api';
import { RootState } from 'redux/modules/root';
import { BuildModel } from 'typings';
import { ThunkAction } from 'redux-thunk';
import { Action } from 'redux';
export const GET_BUILDS_LIST_REQUEST = 'GET_BUILDS_LIST_REQUEST';
export const GET_BUILDS_LIST_SUCCESS = 'GET_BUILDS_LIST_SUCCESS';
export const GET_BUILDS_LIST_FAIL = 'GET_BUILDS_LIST_FAIL';
export const GET_BUILDS_LIST_UPDATE = 'GET_BUILDS_LIST_UPDATE';

export const LOAD_MORE_BUILDS_REQUEST = 'LOAD_MORE_BUILDS_REQUEST';
export const LOAD_MORE_BUILDS_SUCCESS = 'LOAD_MORE_BUILDS_SUCCESS';
export const LOAD_MORE_BUILDS_FAIL = 'LOAD_MORE_BUILDS_FAIL';

export interface GetBuildsListRequestAction {
  type: typeof GET_BUILDS_LIST_REQUEST;
}
const getBuildsListRequest = (): GetBuildsListRequestAction => ({
  type: GET_BUILDS_LIST_REQUEST,
});

export interface GetBuildsListSuccessAction {
  type: typeof GET_BUILDS_LIST_SUCCESS;
  payload: BuildModel[];
}
const getBuildsListSuccess = (payload: BuildModel[]): GetBuildsListSuccessAction => ({
  type: GET_BUILDS_LIST_SUCCESS,
  payload,
});

export interface GetBuildsListFailAction {
  type: typeof GET_BUILDS_LIST_FAIL;
  payload: string;
  error: boolean;
}
const getBuildsListFail = (errorMessage: string): GetBuildsListFailAction => ({
  type: GET_BUILDS_LIST_FAIL,
  payload: errorMessage,
  error: true,
});

function fetchBuildsList(): ThunkAction<void, RootState, unknown, Action<string>> {
  return async (dispatch) => {
    dispatch(getBuildsListRequest());
    try {
      const data = await api.getBuildsList();
      dispatch(getBuildsListSuccess(data));
    } catch (error) {
      dispatch(getBuildsListFail(error.message));
    }
  };
}

function shouldFetchBuildsList(state: RootState) {
  const { builds } = state;

  if (builds.error) {
    return true;
  }

  if (builds?.isFetching || builds?.isLoaded) {
    return false;
  }

  return true;
}

export function fetchBuildsListIfNeeded(): ThunkAction<void, RootState, unknown, Action<string>> {
  return (dispatch, getState) => {
    if (shouldFetchBuildsList(getState())) {
      dispatch(fetchBuildsList());
    }
  };
}

export interface UpdateBuildsListAction {
  type: typeof GET_BUILDS_LIST_UPDATE;
  payload: BuildModel;
}
export const updateBuildsList = (payload: BuildModel): UpdateBuildsListAction => ({
  type: GET_BUILDS_LIST_UPDATE,
  payload,
});

export interface LoadMoreBuildsRequestAction {
  type: typeof LOAD_MORE_BUILDS_REQUEST;
}
const loadMoreBuildsRequest = (): LoadMoreBuildsRequestAction => ({
  type: LOAD_MORE_BUILDS_REQUEST,
});

export interface LoadMoreBuildsSuccessAction {
  type: typeof LOAD_MORE_BUILDS_SUCCESS;
  payload: BuildModel[];
}
const loadMoreBuildsSuccess = (payload: BuildModel[]): LoadMoreBuildsSuccessAction => ({
  type: LOAD_MORE_BUILDS_SUCCESS,
  payload,
});

export interface LoadMoreBuildsFailAction {
  type: typeof LOAD_MORE_BUILDS_FAIL;
  payload: string;
  error: boolean;
}
const loadMoreBuildsFail = (errorMessage: string): LoadMoreBuildsFailAction => ({
  type: LOAD_MORE_BUILDS_FAIL,
  payload: errorMessage,
  error: true,
});

export function fetchMoreBuilds(): ThunkAction<void, RootState, unknown, Action<string>> {
  return async (dispatch, getState) => {
    dispatch(loadMoreBuildsRequest());
    try {
      const { buildsList } = getState().builds;
      const data = await api.getBuildsList(buildsList.length);
      dispatch(loadMoreBuildsSuccess(data));
    } catch (error) {
      dispatch(loadMoreBuildsFail(error.message));
    }
  };
}

export async function addBuild(commitHash: string) {
  return api.addBuild(commitHash);
}

export type BuildsActions =
  | GetBuildsListRequestAction
  | GetBuildsListSuccessAction
  | GetBuildsListFailAction
  | UpdateBuildsListAction
  | LoadMoreBuildsRequestAction
  | LoadMoreBuildsSuccessAction
  | LoadMoreBuildsFailAction;
