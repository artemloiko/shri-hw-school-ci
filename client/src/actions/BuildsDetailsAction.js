import api from '../utils/api';
import { createAction } from '../utils/createAction';

export const INIT_BUILD_DETAILS = 'INIT_BUILD_DETAILS';
export const GET_BUILD_DETAILS_REQUEST = 'GET_BUILD_DETAILS_REQUEST';
export const GET_BUILD_DETAILS_SUCCESS = 'GET_BUILD_DETAILS_SUCCESS';
export const GET_BUILD_DETAILS_FAIL = 'GET_BUILD_DETAILS_FAIL';

export const GET_BUILD_LOGS_REQUEST = 'GET_BUILD_LOGS_REQUEST';
export const GET_BUILD_LOGS_SUCCESS = 'GET_BUILD_LOGS_SUCCESS';
export const GET_BUILD_LOGS_FAIL = 'GET_BUILD_LOGS_FAIL';

const buildDetailsInit = (id) => {
  return createAction(INIT_BUILD_DETAILS, { id });
};

const getBuildDetailsRequest = (id) => {
  return createAction(GET_BUILD_DETAILS_REQUEST, { id });
};
const getBuildDetailsSuccess = (id, data) => {
  return createAction(GET_BUILD_DETAILS_SUCCESS, { id, data });
};
const getBuildDetailsFail = (id, error) => {
  return createAction(GET_BUILD_DETAILS_FAIL, { id, error }, true);
};

function fetchBuildDetailsIfNeeded(buildId) {
  return async (dispatch, getState) => {
    const { buildsDetails } = getState();
    const { details } = buildsDetails[buildId];
    if (details?.isFetching || details?.isLoaded) {
      return;
    }

    dispatch(getBuildDetailsRequest(buildId));
    try {
      const data = await api.getBuildDetails(buildId);
      dispatch(getBuildDetailsSuccess(buildId, data));
    } catch (error) {
      dispatch(getBuildDetailsFail(buildId, error));
    }
  };
}

const getBuildLogsRequest = (id) => {
  return createAction(GET_BUILD_LOGS_REQUEST, { id });
};
const getBuildLogsSuccess = (id, data) => {
  return createAction(GET_BUILD_LOGS_SUCCESS, { id, data });
};
const getBuildLogsFail = (id, error) => {
  return createAction(GET_BUILD_LOGS_FAIL, { id, error }, true);
};

function fetchBuildLogsIfNeeded(buildId) {
  return async (dispatch, getState) => {
    const { buildsDetails } = getState();
    const { logs, details } = buildsDetails[buildId];
    if (logs?.isFetching || logs?.isLoaded) {
      return;
    }

    if (details.status === 'Waiting' || details.status === 'InProgress') {
      dispatch(getBuildLogsSuccess(buildId, ''));
      return;
    }

    dispatch(getBuildLogsRequest(buildId));
    try {
      const data = await api.getBuildLog(buildId);
      dispatch(getBuildLogsSuccess(buildId, data));
    } catch (error) {
      dispatch(getBuildLogsFail(buildId, error));
    }
  };
}

function findLocalBuildDetails(buildId, state) {
  const { builds } = state;
  const { buildsList = [] } = builds;
  return buildsList.find((build) => build.id === buildId);
}

export function getBuildDetails(buildId) {
  return async (dispatch, getState) => {
    const state = getState();

    const { buildsDetails } = state;
    if (!buildsDetails[buildId]) {
      dispatch(buildDetailsInit(buildId));
    }

    const currentBuild = findLocalBuildDetails(buildId, state);
    if (currentBuild) {
      dispatch(getBuildDetailsSuccess(buildId, currentBuild));
    }

    return Promise.all([
      dispatch(fetchBuildDetailsIfNeeded(buildId)),
      dispatch(fetchBuildLogsIfNeeded(buildId)),
    ]);
  };
}
