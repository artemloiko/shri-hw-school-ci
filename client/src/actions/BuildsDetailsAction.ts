import api from '../utils/api';
import { RootState } from 'reducers';
import { BuildModel } from 'typings';
import { ThunkAction } from 'redux-thunk';
import { Action } from 'redux';

export const INIT_BUILD_DETAILS = 'INIT_BUILD_DETAILS';
export const GET_BUILD_DETAILS_REQUEST = 'GET_BUILD_DETAILS_REQUEST';
export const GET_BUILD_DETAILS_SUCCESS = 'GET_BUILD_DETAILS_SUCCESS';
export const GET_BUILD_DETAILS_FAIL = 'GET_BUILD_DETAILS_FAIL';

export const GET_BUILD_LOGS_REQUEST = 'GET_BUILD_LOGS_REQUEST';
export const GET_BUILD_LOGS_SUCCESS = 'GET_BUILD_LOGS_SUCCESS';
export const GET_BUILD_LOGS_FAIL = 'GET_BUILD_LOGS_FAIL';

interface BuildDetailsPayload {
  id: string;
}

export interface BuildDetailsInitAction {
  type: typeof INIT_BUILD_DETAILS;
  payload: BuildDetailsPayload;
}
const buildDetailsInit = (id: string): BuildDetailsInitAction => ({
  type: INIT_BUILD_DETAILS,
  payload: { id },
});

export interface GetBuildDetailsRequestAction {
  type: typeof GET_BUILD_DETAILS_REQUEST;
  payload: BuildDetailsPayload;
}
const getBuildDetailsRequest = (id: string): GetBuildDetailsRequestAction => ({
  type: GET_BUILD_DETAILS_REQUEST,
  payload: { id },
});

export interface GetBuildDetailsSuccessAction {
  type: typeof GET_BUILD_DETAILS_SUCCESS;
  payload: BuildDetailsPayload & { data: BuildModel };
}
const getBuildDetailsSuccess = (id: string, data: BuildModel): GetBuildDetailsSuccessAction => ({
  type: GET_BUILD_DETAILS_SUCCESS,
  payload: { id, data },
});

export interface GetBuildDetailsFailAction {
  type: typeof GET_BUILD_DETAILS_FAIL;
  payload: BuildDetailsPayload & { error: string };
  error: boolean;
}
const getBuildDetailsFail = (id: string, error: string): GetBuildDetailsFailAction => ({
  type: GET_BUILD_DETAILS_FAIL,
  payload: { id, error },
  error: true,
});

function fetchBuildDetailsIfNeeded(
  buildId: string,
): ThunkAction<void, RootState, unknown, Action<string>> {
  return async (dispatch, getState) => {
    const { buildsDetails } = getState();
    const { details } = buildsDetails[buildId];
    if (details?.isFetching || details?.isLoaded) {
      return;
    }

    dispatch(getBuildDetailsRequest(buildId));
    try {
      const data = await api.getBuildDetails(buildId);
      dispatch(getBuildDetailsSuccess(buildId, data));
    } catch (error) {
      dispatch(getBuildDetailsFail(buildId, error.message));
    }
  };
}

export interface GetBuildLogsRequestAction {
  type: typeof GET_BUILD_LOGS_REQUEST;
  payload: BuildDetailsPayload;
}
const getBuildLogsRequest = (id: string): GetBuildLogsRequestAction => ({
  type: GET_BUILD_LOGS_REQUEST,
  payload: { id },
});

export interface GetBuildLogsSuccessAction {
  type: typeof GET_BUILD_LOGS_SUCCESS;
  payload: BuildDetailsPayload & { data: string };
}
const getBuildLogsSuccess = (id: string, data: string): GetBuildLogsSuccessAction => ({
  type: GET_BUILD_LOGS_SUCCESS,
  payload: { id, data },
});

export interface GetBuildLogsFailAction {
  type: typeof GET_BUILD_LOGS_FAIL;
  payload: BuildDetailsPayload & { error: string };
  error: boolean;
}
const getBuildLogsFail = (id: string, error: string): GetBuildLogsFailAction => ({
  type: GET_BUILD_LOGS_FAIL,
  payload: { id, error },
  error: true,
});

function fetchBuildLogsIfNeeded(
  buildId: string,
): ThunkAction<void, RootState, unknown, Action<string>> {
  return async (dispatch, getState) => {
    const { buildsDetails } = getState();
    const { logs, details } = buildsDetails[buildId];
    if (logs?.isFetching || logs?.isLoaded) {
      return;
    }

    if (details?.data?.status === 'Waiting' || details?.data?.status === 'InProgress') {
      dispatch(getBuildLogsSuccess(buildId, ''));
      return;
    }

    dispatch(getBuildLogsRequest(buildId));
    try {
      const data = await api.getBuildLog(buildId);
      dispatch(getBuildLogsSuccess(buildId, data));
    } catch (error) {
      dispatch(getBuildLogsFail(buildId, error.message));
    }
  };
}

function findLocalBuildDetails(buildId: string, state: RootState): BuildModel | undefined {
  const { builds } = state;
  const { buildsList = [] } = builds;
  return buildsList.find((build) => build.id === buildId);
}

export function getBuildDetails(
  buildId: string,
): ThunkAction<void, RootState, unknown, Action<string>> {
  return async (dispatch, getState) => {
    const state = getState();

    const { buildsDetails } = state;
    if (!buildsDetails[buildId]) {
      dispatch(buildDetailsInit(buildId));
    }

    const currentBuild = findLocalBuildDetails(buildId, state);
    if (currentBuild) {
      dispatch(getBuildDetailsSuccess(buildId, currentBuild));
    }

    return Promise.all([
      dispatch(fetchBuildDetailsIfNeeded(buildId)),
      dispatch(fetchBuildLogsIfNeeded(buildId)),
    ]);
  };
}

export type BuildDetailsActions =
  | BuildDetailsInitAction
  | GetBuildDetailsRequestAction
  | GetBuildDetailsSuccessAction
  | GetBuildDetailsFailAction
  | GetBuildLogsRequestAction
  | GetBuildLogsSuccessAction
  | GetBuildLogsFailAction;
