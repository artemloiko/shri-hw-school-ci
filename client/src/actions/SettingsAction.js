import { createAction } from '../utils/createAction';
export const GET_SETTINGS_REQUEST = 'GET_SETTINGS_REQUEST';
export const GET_SETTINGS_SUCCESS = 'GET_SETTINGS_SUCCESS';
export const GET_SETTINGS_FAIL = 'GET_SETTINGS_FAIL';

const getSettingsRequest = () => {
  return createAction(GET_SETTINGS_REQUEST);
};

const getSettingsSuccess = (data) => {
  return createAction(GET_SETTINGS_SUCCESS, data);
};

const getSettingsFail = (error) => {
  return createAction(GET_SETTINGS_FAIL, error, true);
};

async function mockApi() {
  await new Promise((resolve, reject) => {
    setTimeout(resolve, 1500);
  });
  return {
    id: 'edba260c-e190-48b6-8ff5-1fac2778f022',
    repoName: 'artuom130/isItReal',
    buildCommand: 'npm run build',
    mainBranch: 'master',
    period: 0,
  };
}

function fetchSettings() {
  return async (dispatch) => {
    dispatch(getSettingsRequest());
    try {
      const data = await mockApi();
      dispatch(getSettingsSuccess(data));
    } catch (error) {
      dispatch(getSettingsFail(error));
    }
  };
}

function shouldFetchSettings(state) {
  const { settings } = state;
  if (!settings.id) {
    return true;
  }
  return false;
}

export function fetchSettingsIfNeeded() {
  return (dispatch, getState) => {
    if (shouldFetchSettings(getState())) {
      dispatch(fetchSettings());
    }
  };
}
