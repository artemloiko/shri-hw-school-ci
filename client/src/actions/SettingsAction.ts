import api from '../utils/api';
import { ConfigurationModel, ConfigurationDTO } from 'typings';
import { RootState } from 'reducers';
import { ThunkAction } from 'redux-thunk';
import { Action } from 'redux';

export const GET_SETTINGS_REQUEST = 'GET_SETTINGS_REQUEST';
export const GET_SETTINGS_SUCCESS = 'GET_SETTINGS_SUCCESS';
export const GET_SETTINGS_FAIL = 'GET_SETTINGS_FAIL';
export const SET_SETTINGS = 'SET_SETTINGS';
export const SET_SETTINGS_FAIL = 'SET_SETTINGS_FAIL';
export const RESET_SETTINGS_ERROR = 'RESET_SETTINGS_ERROR';

export interface GetSettingsRequestAction {
  type: typeof GET_SETTINGS_REQUEST;
}
const getSettingsRequest = (): GetSettingsRequestAction => ({
  type: GET_SETTINGS_REQUEST,
});

export interface GetSettingsSuccessAction {
  type: typeof GET_SETTINGS_SUCCESS;
  payload: ConfigurationModel;
}
const getSettingsSuccess = (data: ConfigurationModel): GetSettingsSuccessAction => ({
  type: GET_SETTINGS_SUCCESS,
  payload: data,
});

export interface GetSettingsFailAction {
  type: typeof GET_SETTINGS_FAIL;
  payload: string;
  error: boolean;
}
const getSettingsFail = (errorMessage: string): GetSettingsFailAction => ({
  type: GET_SETTINGS_FAIL,
  payload: errorMessage,
  error: true,
});

function fetchSettings(): ThunkAction<void, RootState, unknown, Action<string>> {
  return async (dispatch) => {
    dispatch(getSettingsRequest());
    try {
      const data = await api.getSettings();
      dispatch(getSettingsSuccess(data));
    } catch (error) {
      dispatch(getSettingsFail(error.message));
    }
  };
}

function shouldFetchSettings(state: RootState) {
  const { settings } = state;

  if (settings?.isFetching || settings?.isLoaded) {
    return false;
  }

  return true;
}

export function fetchSettingsIfNeeded(): ThunkAction<void, RootState, unknown, Action<string>> {
  return (dispatch, getState) => {
    if (shouldFetchSettings(getState())) {
      dispatch(fetchSettings());
    }
  };
}
export interface UpdateSettingsAction {
  type: typeof SET_SETTINGS;
  payload: ConfigurationDTO;
}
export function updateSettings(settingsDTO: ConfigurationDTO): UpdateSettingsAction {
  return { type: SET_SETTINGS, payload: settingsDTO };
}

export interface UpdateSettingsFailAction {
  type: typeof SET_SETTINGS_FAIL;
  payload: string;
  error: boolean;
}
export function updateSettingsFail(errorMessage: string): UpdateSettingsFailAction {
  return { type: SET_SETTINGS_FAIL, payload: errorMessage, error: true };
}

export interface ResetSettingsErrorAction {
  type: typeof RESET_SETTINGS_ERROR;
}
export function resetSettingsError(): ResetSettingsErrorAction {
  return { type: RESET_SETTINGS_ERROR };
}

export type SettingsActions =
  | GetSettingsRequestAction
  | GetSettingsSuccessAction
  | GetSettingsFailAction
  | UpdateSettingsAction
  | UpdateSettingsFailAction
  | ResetSettingsErrorAction;
