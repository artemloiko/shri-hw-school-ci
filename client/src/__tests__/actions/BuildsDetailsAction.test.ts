import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mocked } from 'ts-jest/utils';
import { getBuildDetails } from '../../actions/BuildsDetailsAction';
import apiMock from '../../utils/api';
import { BuildModel } from 'typings';

jest.mock('../../utils/api');

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const buildId = '30c10a6a-087e-4a6b-aed8-8a809169a305';
const initialState = {
  builds: {},
  buildsDetails: {
    [buildId]: {
      details: { isFetching: false, isLoaded: false },
      logs: { isFetching: false, isLoaded: false },
    },
  },
};

describe('builds details actions', () => {
  test('fetching builds details success', async () => {
    const store = mockStore(initialState);
    const details: BuildModel = {
      id: buildId,
      configurationId: 'e4a75284-3cbb-4676-8db6-4e3012958328',
      buildNumber: 1,
      commitMessage: 'Add option to work with async fn',
      commitHash: 'a63f2b3',
      branchName: 'master',
      authorName: 'Artem Loiko',
      status: 'Success',
      start: '2020-04-10T07:07:14.337',
      duration: 3254,
    };
    const logs = '\u001b[32mCompiled successfully.\u001b[39m\u001b[32m\u001b[39m';
    mocked(apiMock.getBuildDetails).mockImplementationOnce(() => Promise.resolve(details));
    mocked(apiMock.getBuildLog).mockImplementationOnce(() => Promise.resolve(logs));

    await store.dispatch<any>(getBuildDetails(buildId));

    expect(store.getActions()).toMatchSnapshot();
  });

  test('fetching builds details fail', async () => {
    const store = mockStore(initialState);
    const error = {
      message: 'Request failed with status code 500',
    };
    mocked(apiMock.getBuildDetails).mockImplementationOnce(() => Promise.reject(error));
    mocked(apiMock.getBuildLog).mockImplementationOnce(() => Promise.reject(error));

    await store.dispatch<any>(getBuildDetails(buildId));

    expect(store.getActions()).toMatchSnapshot();
  });
});
