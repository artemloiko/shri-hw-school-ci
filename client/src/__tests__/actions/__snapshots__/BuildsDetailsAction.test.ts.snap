// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds details actions fetching builds details fail 1`] = `
Array [
  Object {
    "payload": Object {
      "id": "30c10a6a-087e-4a6b-aed8-8a809169a305",
    },
    "type": "GET_BUILD_DETAILS_REQUEST",
  },
  Object {
    "payload": Object {
      "id": "30c10a6a-087e-4a6b-aed8-8a809169a305",
    },
    "type": "GET_BUILD_LOGS_REQUEST",
  },
  Object {
    "error": true,
    "payload": Object {
      "error": "Request failed with status code 500",
      "id": "30c10a6a-087e-4a6b-aed8-8a809169a305",
    },
    "type": "GET_BUILD_DETAILS_FAIL",
  },
  Object {
    "error": true,
    "payload": Object {
      "error": "Request failed with status code 500",
      "id": "30c10a6a-087e-4a6b-aed8-8a809169a305",
    },
    "type": "GET_BUILD_LOGS_FAIL",
  },
]
`;

exports[`builds details actions fetching builds details success 1`] = `
Array [
  Object {
    "payload": Object {
      "id": "30c10a6a-087e-4a6b-aed8-8a809169a305",
    },
    "type": "GET_BUILD_DETAILS_REQUEST",
  },
  Object {
    "payload": Object {
      "id": "30c10a6a-087e-4a6b-aed8-8a809169a305",
    },
    "type": "GET_BUILD_LOGS_REQUEST",
  },
  Object {
    "payload": Object {
      "data": Object {
        "authorName": "Artem Loiko",
        "branchName": "master",
        "buildNumber": 1,
        "commitHash": "a63f2b3",
        "commitMessage": "Add option to work with async fn",
        "configurationId": "e4a75284-3cbb-4676-8db6-4e3012958328",
        "duration": 3254,
        "id": "30c10a6a-087e-4a6b-aed8-8a809169a305",
        "start": "2020-04-10T07:07:14.337",
        "status": "Success",
      },
      "id": "30c10a6a-087e-4a6b-aed8-8a809169a305",
    },
    "type": "GET_BUILD_DETAILS_SUCCESS",
  },
  Object {
    "payload": Object {
      "data": "[32mCompiled successfully.[39m[32m[39m",
      "id": "30c10a6a-087e-4a6b-aed8-8a809169a305",
    },
    "type": "GET_BUILD_LOGS_SUCCESS",
  },
]
`;
